### ISSUE TITLE
azcopy crashes during sync or copy of a whole Premium FileStorage to one another

### ISSUE THREAD

**Issue by flootsk** [Azure/azure-storage-azcopy#1555]
##### Note: The version is visible when running AzCopy without any argument
10.12.1

### Which platform are you using? (ex: Windows, Mac, Linux)
Linux, Windows

##### Note: Please remove the SAS to avoid exposing your credentials. If you cannot remember the exact command, please retrieve it from the beginning of the log file.
```
./azcopy sync 'https://<sourcestorageaccount>.file.core.windows.net/<fileshare>?<SAS>' 'https://<deststorageaccount>.file.core.windows.net/<fileshare>?<SAS>' --log-level INFO --recursive=true
```

### What problem was encountered?
`azcopy` crashes during the scan phase after a random number of files scanned (e.g. sometime 2000, sometime 60000). It also happens during the copy with `azcopy copy [...]`

Here is a sample of what I get on stderr:

```
fatal error: concurrent map read and map write

goroutine 70 [running]:
runtime.throw(0xf70464, 0x21)
        /opt/hostedtoolcache/go/1.16.0/x64/src/runtime/panic.go:1117 +0x72 fp=0xc013f0dc40 sp=0xc013f0dc10 pc=0x438fd2
runtime.mapaccess2_faststr(0xe12200, 0xc0119b1c20, 0xc002149100, 0x7f, 0xc002149100, 0xc013f0dcd8)
        /opt/hostedtoolcache/go/1.16.0/x64/src/runtime/map_faststr.go:116 +0x4a5 fp=0xc013f0dcb0 sp=0xc013f0dc40 pc=0x415e85
github.com/Azure/azure-storage-azcopy/v10/ste.(*jpptFolderTracker).RecordCreation(0xc0116bc920, 0xc002149100, 0x7f)
        /home/vsts/work/1/s/ste/folderCreationTracker.go:80 +0x94 fp=0xc013f0dd08 sp=0xc013f0dcb0 pc=0xc7a554
github.com/Azure/azure-storage-azcopy/v10/ste.AzureFileParentDirCreator.CreateDirToRoot(0x10ca038, 0xc013c682a0, 0xc008d1da00, 0x5, 0x0, 0x0, 0x0, 0xc008d1da08, 0x2c, 0xc008d1da34, ...)
        /home/vsts/work/1/s/ste/sender-azureFile.go:455 +0xa55 fp=0xc013f0f728 sp=0xc013f0dd08 pc=0xc9a9f5
github.com/Azure/azure-storage-azcopy/v10/ste.(*azureFileSenderBase).EnsureFolderExists(0xc00785bba0, 0x0, 0x0)
        /home/vsts/work/1/s/ste/sender-azureFile.go:355 +0x13a fp=0xc013f0f958 sp=0xc013f0f728 pc=0xc990da
github.com/Azure/azure-storage-azcopy/v10/ste.anyToRemote_folder(0x10d94d0, 0xc00110d500, 0xff4eb712c2795bb5, 0x44580b6a9b0a6a44, 0x800000, 0xc008d1db00, 0xf0, 0x0, 0xc008d1da00, 0xf8, ...)
        /home/vsts/work/1/s/ste/xfer-anyToRemote-folder.go:70 +0x2b4 fp=0xc013f0fae8 sp=0xc013f0f958 pc=0xcb0354
github.com/Azure/azure-storage-azcopy/v10/ste.anyToRemote(0x10d94d0, 0xc00110d500, 0x10b7380, 0xc015508c00, 0x7f78c0a6a088, 0xc000418000, 0xfbaa70, 0xfbaa30)
        /home/vsts/work/1/s/ste/xfer-anyToRemote-file.go:179 +0x198 fp=0xc013f0fe40 sp=0xc013f0fae8 pc=0xcac3f8
github.com/Azure/azure-storage-azcopy/v10/ste.parameterizeSend.func1(0x10d94d0, 0xc00110d500, 0x10b7380, 0xc015508c00, 0x7f78c0a6a088, 0xc000418000)
        /home/vsts/work/1/s/ste/xfer.go:76 +0x73 fp=0xc013f0fe90 sp=0xc013f0fe40 pc=0xcc5b13
github.com/Azure/azure-storage-azcopy/v10/ste.(*jobPartMgr).StartJobXfer(0xc015dafc00, 0x10d94d0, 0xc00110d500)
        /home/vsts/work/1/s/ste/mgr-JobPartMgr.go:676 +0x71 fp=0xc013f0fed0 sp=0xc013f0fe90 pc=0xc883d1
github.com/Azure/azure-storage-azcopy/v10/ste.(*jobPartTransferMgr).StartJobXfer(0xc00110d500)
        /home/vsts/work/1/s/ste/mgr-JobPartTransferMgr.go:215 +0x48 fp=0xc013f0fef8 sp=0xc013f0fed0 pc=0xc89648
github.com/Azure/azure-storage-azcopy/v10/ste.(*jobsAdmin).transferProcessor.func1(0x10d94d0, 0xc00110d500)
        /home/vsts/work/1/s/ste/JobsAdmin.go:465 +0x1b7 fp=0xc013f0ff78 sp=0xc013f0fef8 pc=0xcb8bf7
github.com/Azure/azure-storage-azcopy/v10/ste.(*jobsAdmin).transferProcessor(0xc00041a000, 0x2e)
        /home/vsts/work/1/s/ste/JobsAdmin.go:473 +0x8f fp=0xc013f0ffd0 sp=0xc013f0ff78 pc=0xc74eaf
runtime.goexit()
        /opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc013f0ffd8 sp=0xc013f0ffd0 pc=0x470aa1
created by github.com/Azure/azure-storage-azcopy/v10/ste.initJobsAdmin
        /home/vsts/work/1/s/ste/JobsAdmin.go:219 +0x7bd

goroutine 1 [chan receive]:
github.com/Azure/azure-storage-azcopy/v10/cmd.(*fileTraverser).traverse(0xc000154200, 0x0, 0xc0003eb2b0, 0x16b3d50, 0x0, 0x0, 0x0, 0x0)
        /home/vsts/work/1/s/cmd/zc_traverser_file.go:236 +0x6aa
github.com/Azure/azure-storage-azcopy/v10/cmd.(*syncEnumerator).enumerate(0xc000090550, 0x10ca000, 0xc0003e4870)
        /home/vsts/work/1/s/cmd/zc_enumerator.go:631 +0xf4
github.com/Azure/azure-storage-azcopy/v10/cmd.(*cookedSyncCmdArgs).process(0xc000248e00, 0x13f, 0x0)
        /home/vsts/work/1/s/cmd/sync.go:678 +0x3e5
github.com/Azure/azure-storage-azcopy/v10/cmd.init.21.func2(0xc000428a00, 0xc0001014a0, 0x2, 0x5)
        /home/vsts/work/1/s/cmd/sync.go:713 +0x17a
github.com/spf13/cobra.(*Command).execute(0xc000428a00, 0xc000101450, 0x5, 0x5, 0xc000428a00, 0xc000101450)
        /home/vsts/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:860 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x1677640, 0x44580b6a9b0a6a44, 0xc0002dfdf8, 0x0)
        /home/vsts/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /home/vsts/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
github.com/Azure/azure-storage-azcopy/v10/cmd.Execute(0xc000176f80, 0x1b, 0xc000176f80, 0x1b, 0xc000150420, 0x21, 0x3ffff)
        /home/vsts/work/1/s/cmd/root.go:165 +0xf8
main.main()
        /home/vsts/work/1/s/main.go:82 +0x3a5

[and so on with many other goroutine blocks]
```

### How can we reproduce the problem in the simplest way?

1. Create 2 Premium FileStorage accounts (1 will be the source storage account, and the other 1 will be the destination), with 1 file share in each
2. In the source storage account, put/generate at least 100k files distributed in a ~6 levels of directory depth
3. Get SAS tokens for the source and destination storage
3. Run `azcopy copy` for the whole source file share to the destination file share, using their respective SAS token

### Have you found a mitigation/solution?

Downgrade to version 10.11.0 of `azcopy`.

**Comment by rorkea** [comment]
Seeing the same errors on the same version.
Distribution of ~30TB of data across 50 distinct fileshares with random sizes but both sync & copy operations fail after some period of time.
The error handling around the goroutine also supercedes any of the logging and thus isn't directly capturable in the log files, or via piping the ps output to a file.

**Comment by Dragonsbane** [comment]
Can confirm this behavior using azcopy copy to transfer data from a local disk to an Azure File Share.

SO: Ubuntu 18.04
Files: 22169
Folders: 7260
TotalBytes: 221MB

azcopy --log-level ERROR copy '/REDACTEDorigem' 'REDACTED-StorageAccountSASURL' --recursive
INFO: Scanning...
INFO: Any empty folders will be processed, because source and destination both support folders

Job 3294d529-98bc-3740-4ce5-50ac3b422b9b has started
Log file is located at: .azcopy/3294d529-98bc-3740-4ce5-50ac3b422b9b.log

0.0 %, 0 Done, 0 Failed, 10000 Pending, 0 Skipped, 10000 Total (scanning...), 
1.8 %, 376 Done, 0 Failed, 29053 Pending, 0 Skipped, 29429 Total, 2-sec Throughput (Mb/s): 17.0232
fatal error: concurrent map read and map write

goroutine 52 [running]:
runtime.throw(0xf7304a, 0x21)
/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/panic.go:1117 +0x72 fp=0xc00244d7d8 sp=0xc00244d7a8 pc=0x438fd2
.......

I´m yet to test the downgrade to version 10.11.0 as suggested by the op.

**Comment by adreed-msft** [comment]
This is pretty wild to hear, since the folder creation tracker has a mutex on it. No two threads _can_ ever be writing to this map. Would one of you mind dropping the full set of stack traces? That should help dramatically.

**Comment by gapra-msft** [comment]
Hi, I believe [this](https://github.com/Azure/azure-storage-azcopy/pull/1585) PR resolved this issue and was released with 10.13.0.

I also attempted to reproduce this issue on the latest version of AzCopy and was not successful. I will go ahead and close this issue. In case you are still experiencing it, please feel free to reopen the issue.
